cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_heap_dpc.cpp

```diff
@@ -0,0 +1,471 @@
+
+    using acc_t =
+        sycl::accessor<sel_t, 1, sycl::access::mode::read_write, sycl::access::target::local>;
+
+    using dimension_keeper_t = dimension_keeper<Float, dst_out, ids_out>;
+
+public:
+    kernel_select_heap() = default;
+
+    template <bool ids_only = (ids_out && !dst_out), typename = std::enable_if_t<ids_only>>
+    kernel_select_heap(const dst_t* const src_ptr,
+                       idx_t* const ids_ptr,
+                       std::int32_t k,
+                       std::int32_t width,
+                       std::int32_t height,
+                       std::int32_t src_str,
+                       std::int32_t ids_str,
+                       acc_t heaps)
+            : dk_(ids_str, ids_ptr),
+              src_str_(src_str),
+              src_ptr_(src_ptr),
+              k_(k),
+              width_(width),
+              height_(height),
+              heaps_(std::move(heaps)) {}
+
+    template <bool dst_only = (dst_out && !ids_out), typename = std::enable_if_t<dst_only>>
+    kernel_select_heap(const dst_t* const src_ptr,
+                       dst_t* const dst_ptr,
+                       std::int32_t k,
+                       std::int32_t width,
+                       std::int32_t height,
+                       std::int32_t src_str,
+                       std::int32_t dst_str,
+                       acc_t heaps)
+            : dk_(dst_str, dst_ptr),
+              src_str_(src_str),
+              src_ptr_(src_ptr),
+              k_(k),
+              width_(width),
+              height_(height),
+              heaps_(std::move(heaps)) {}
+
+    template <bool both = (ids_out && dst_out), typename = std::enable_if_t<both>>
+    kernel_select_heap(const dst_t* const src_ptr,
+                       idx_t* const ids_ptr,
+                       dst_t* const dst_ptr,
+                       std::int32_t k,
+                       std::int32_t width,
+                       std::int32_t height,
+                       std::int32_t src_str,
+                       std::int32_t ids_str,
+                       std::int32_t dst_str,
+                       acc_t heaps)
+            : dk_(ids_str, dst_str, ids_ptr, dst_ptr),
+              src_str_(src_str),
+              src_ptr_(src_ptr),
+              k_(k),
+              width_(width),
+              height_(height),
+              heaps_(std::move(heaps)) {}
+
+    void operator()(sycl::nd_item<1> item) const {
+        auto sg = item.get_sub_group();
+
+        const std::int32_t sid = sg.get_group_linear_id();
+        const std::int32_t cid = sg.get_local_linear_id();
+        const std::int32_t sg_width = sg.get_local_range().size();
+        const std::int32_t wg_width = item.get_local_range().size();
+        const std::int32_t rid = sid + item.get_group_linear_id() * sg.get_group_range().size();
+
+        // Check for the case if sg_size != sg_width
+        if (sid * proposed_sg_size >= wg_width)
```

Comment:
For SYCL it's not an option, right?

Reply:
As you may know there is no way to really force compiler to choose specific sub group size